### ISSUE TITLE
Not obtaining consistent results

### ISSUE THREAD

**Issue by AlexisPister** [Accenture/AmpliGraph#92]
### Description
Hi,

I made a jupyter notebook where I load several n3 files with rdflib, filter some triplets and then transform each graph into a numpy array for ampligraph. I then concatenante all graph and create a ampligraph model and test it with MRR, MR and hits10.

However, I don't always get the same results. I put seeds where I could, and I start the notebook with random.seed(0) and np.random.seed(0).

The numpy array describing the input graph for the ampligraph model seems to always be in the same order, but the results differs. 

### Steps to Reproduce

```
random.seed(0)
np.random.seed(0)

# Remove triplets with literals of unrelevant information from rdflib graph
def remove_triplets(rdflib_graph):
    new_g = Graph()
    
    for s,p,o in rdflib_graph:
        if type(s) != rdflib.term.Literal and type(p) != rdflib.term.Literal and type(o) != rdflib.term.Literal:
            if p not in TO_REMOVE:
                new_g.add((s, p, o))
            
    return new_g

# From path, process graph and return a numpy ampligraph object ready for embedding task
def process(path):
    # Load rdflibj graph from path
    rdflibg = Graph()
    data = open(path, "rb")
    rdflibg.parse(data, format="n3")
    
    # Remove unecessary triplets
    rdflibg = remove_triplets(rdflibg)
    amplg = np.array(rdflibg)
    return amplg

agrelong = process("agrelon_2018-10-16.n3")
biog = process("bio_2011-06-14.n3")
cong = process("con_2011-02-01.n3")
duvg = process("duv_2016-08-30.n3")
emotiong = process("emotion_2013-11-26.n3")

amplg = np.vstack((agrelong, biog, cong, duvg, emotiong))

# Divide between train, test and valid set
train_validg, testg = train_test_split_no_unseen(amplg, 200, seed=0)
traing, validg = train_test_split_no_unseen(train_validg, 200, seed=0)

data = {"train":traing, "test":testg, "valid":validg}


# Train an embedding model and save it
def init_model(model_name, nbatches, epochs, k, eta, optimizer='adam'):
    model = model_name(batches_count=nbatches, seed=0, epochs=epochs, k=k, eta=eta,
                    # Use adam optimizer with learning rate 1e-3
                    optimizer='adam', optimizer_params={'lr':1e-3},
                    # Use pairwise loss with margin 0.5
                    loss='pairwise', loss_params={'margin':0.5},
                    # Use L2 regularizer with regularizer weight 1e-5
                    regularizer='LP', regularizer_params={'p':2, 'lambda':1e-5}, 
                    # Enable stdout messages (set to false if you don't want to display)
                    verbose=False)
    return model

# Train an ampligraph model and return it fitted
def training(model, train_data, test_data, valid_data, early_stop=False):
    #Training
    np.random.seed(0)
    random.seed(0)
    if early_stop == False:
        model.fit(train_data)
        return model
    else:
        print("early stop heuristic")
        model.fit(train_data,
                    early_stopping = True,
                    early_stopping_params = {
                              'x_valid': valid_data,       # validation set
                              'criteria':'hits10',         # Uses hits10 criteria for early stopping
                              'burn_in': 100,              # early stopping kicks in after 100 epochs
                              'check_interval':2,          # validates every 20th epoch
                              'stop_interval':5,       # stops if 5 successive validation checks are bad.
                              'x_filter': np.concatenate((train_data, test_data, valid_data)),          # Use filter for filtering out positives 
                              'corruption_entities': "all", # corrupt using all entities
                              'corrupt_side':'s+o'         # corrupt subject and object (but not at once)
                          }
             )
    return model

model = init_model(ComplEx, nbatches=10, epochs=100, k=100, eta=10)
model = training(model, traing, testg, validg, early_stop=True)

# Evaluation protocol
ranks = evaluate_performance(testg, 
                                 model=model, 
                                 use_default_protocol=True, # corrupt subj and obj separately while evaluating
                                 filter_triples=amplg,
                                 verbose=True)


mrr = mrr_score(ranks)
mr = mr_score(ranks)
hits_10 = hits_at_n_score(ranks, n=10)
print("MRR: %f, MR: %f, Hits@10: %f" % (mrr, mr, hits_10))

```

When I run this notebook from scratch, I don't always get the same metrics. Do you have any idea where this could come from ?

Thanks

**Comment by gatemezing** [comment]
I can also add the results of the notebook behave differently from one operating system to another.

**Comment by sumitpai** [comment]
@AlexisPister Can you please elaborate on how different the metrics are. I tried running the second part of your code with wn18 dataset in an ipython notebook. I was able to get same results. Below is the snippet that I tried.

```
import random
from rdflib import Graph
import numpy as np
import os
random.seed(0)
np.random.seed(0)

from ampligraph.datasets import  load_wn18
from ampligraph.evaluation import hits_at_n_score, mrr_score, mr_score
from ampligraph.latent_features import ComplEx, DistMult, HolE
from ampligraph.evaluation import evaluate_performance


# Train an embedding model and save it
def init_model(model_name, nbatches, epochs, k, eta, optimizer='adam'):
    model = model_name(batches_count=nbatches, seed=0, epochs=epochs, k=k, eta=eta,
                    # Use adam optimizer with learning rate 1e-3
                    optimizer='adam', optimizer_params={'lr':1e-3},
                    # Use pairwise loss with margin 0.5
                    loss='pairwise', loss_params={'margin':0.5},
                    # Use L2 regularizer with regularizer weight 1e-5
                    regularizer='LP', regularizer_params={'p':2, 'lambda':1e-5}, 
                    # Enable stdout messages (set to false if you don't want to display)
                    verbose=True)
    return model

# Train an ampligraph model and return it fitted
def training(model, train_data, test_data, valid_data, early_stop=False):
    #Training
    np.random.seed(0)
    random.seed(0)
    if early_stop == False:
        model.fit(train_data)
        return model
    else:
        print("early stop heuristic")
        model.fit(train_data,
                    early_stopping = True,
                    early_stopping_params = {
                              'x_valid': valid_data,       # validation set
                              'criteria':'hits10',         # Uses hits10 criteria for early stopping
                              'burn_in': 100,              # early stopping kicks in after 100 epochs
                              'check_interval':2,          # validates every 20th epoch
                              'stop_interval':5,       # stops if 5 successive validation checks are bad.
                              'x_filter': np.concatenate((train_data, test_data, valid_data)),          # Use filter for filtering out positives 
                              'corruption_entities': "all", # corrupt using all entities
                              'corrupt_side':'s+o'         # corrupt subject and object (but not at once)
                          }
             )
    return model
X = load_wn18()
traing = X["train"]
testg = X["test"]
validg = X["valid"]
model = init_model(ComplEx, nbatches=10, epochs=100, k=100, eta=10)

model = training(model, traing, testg, validg, early_stop=True)

amplg = np.concatenate((traing, testg, validg), 0)
# Evaluation protocol
ranks = evaluate_performance(testg, 
                                 model=model, 
                                 use_default_protocol=True, # corrupt subj and obj separately while evaluating
                                 filter_triples=amplg,
                                 verbose=True)


mrr = mrr_score(ranks)
mr = mr_score(ranks)
hits_10 = hits_at_n_score(ranks, n=10)
print("MRR: %f, MR: %f, Hits@10: %f" % (mrr, mr, hits_10))
```
Training loss after 100 epochs is :
```
Average Loss:   0.000288:  99%|█████████▉| 99/100 [01:05<00:00,  1.54epoch/s]
```

Output metric is:
```
MRR: 0.891770, MR: 677.525500, Hits@10: 0.935900
```

I also tested the train_test_split_no_unseen function as follows (by splitting the train set):
```
from ampligraph.evaluation import train_test_split_no_unseen
train_validg, testg = train_test_split_no_unseen(traing, 200, seed=0)
train_validg2, testg2 = train_test_split_no_unseen(traing, 200, seed=0)
np.testing.assert_array_equal(train_validg,train_validg2)
np.testing.assert_array_equal(testg,testg2)
print(train_validg[:5], testg[:5])
print(train_validg2[:5], testg2[:5])
```
To be sure, manually checked a few entries. Even here the output was reproducible:
```
[['03964744' '_hyponym' '04371774']
 ['00260881' '_hypernym' '00260622']
 ['02199712' '_member_holonym' '02188065']
 ['01332730' '_derivationally_related_form' '03122748']
 ['06066555' '_derivationally_related_form' '00645415']] [['07309781' '_hyponym' '07351909']
 ['09818343' '_derivationally_related_form' '02133297']
 ['12794568' '_member_holonym' '12792638']
 ['14416845' '_derivationally_related_form' '00714273']
 ['02303331' '_derivationally_related_form' '00205891']]
[['03964744' '_hyponym' '04371774']
 ['00260881' '_hypernym' '00260622']
 ['02199712' '_member_holonym' '02188065']
 ['01332730' '_derivationally_related_form' '03122748']
 ['06066555' '_derivationally_related_form' '00645415']] [['07309781' '_hyponym' '07351909']
 ['09818343' '_derivationally_related_form' '02133297']
 ['12794568' '_member_holonym' '12792638']
 ['14416845' '_derivationally_related_form' '00714273']
 ['02303331' '_derivationally_related_form' '00205891']]
```

---------------------------------------

Looking at the results on wn18, my assumption is that the process method in the attached snippet might not be producing reproducible results (i.e. the graph output of the process method might be same but the ordering of the triples might be different). I assume that this is getting propagated, and the train -test split is giving different splits everytime which is getting reflected in the metrics. 

As a test, I would request you to verify the `amplg` variable to see if it is deterministic. Can you please verify that. You can also do the following and see if the assertion works:
```
agrelong = process("agrelon_2018-10-16.n3")
biog = process("bio_2011-06-14.n3")
cong = process("con_2011-02-01.n3")
duvg = process("duv_2016-08-30.n3")
emotiong = process("emotion_2013-11-26.n3")
amplg = np.vstack((agrelong, biog, cong, duvg, emotiong))

agrelong2 = process("agrelon_2018-10-16.n3")
biog2 = process("bio_2011-06-14.n3")
cong2 = process("con_2011-02-01.n3")
duvg2= process("duv_2016-08-30.n3")
emotiong2 = process("emotion_2013-11-26.n3")
amplg2 = np.vstack((agrelong2, biog2, cong2, duvg2, emotiong2))
np.testing.assert_array_equal(amplg, amplg2)

print(amplg[:5])

```

In addition to that, if you are using GPU can you specify the CUDA version(```nvcc --version```), the NVIDIA Driver version (listed in ```nvidia-smi```) and the TensorFlow version(```tf.__version__```) that is being used.

**Comment by sumitpai** [comment]
rdflib docs mentions that an RDF Graph is a *set of RDF triples*. And I believe that this set is an unordered set, due to which the for loop which iterates over the RDF Graph (in the ```remove_triplets``` method) is not deterministic.

**Comment by AlexisPister** [comment]
Thanks for the feedback !

Indeed, if I call the process function 2 times, the returned objects are differents. I thought it alwyas gave the same results because when I printed it, the beginning of the array was not changing over time. To always have the same graph, I thought of first sorting the array and then suffling it with a determined seed. However, it seems that the ids of the blank nodes change all the times when the rdflib graph is loaded, so the order of the graph will always be different.

I guess I should do the preprocess step always one time for my graphs and then save the results in .n3 files to avoid the stochasticity of the preprocess.
